docs/guides/contexts_expressions.md

```diff
@@ -0,0 +1,390 @@
+Out[9]: True
+```
+
+and it can also be combined with other expressions & constants using
+**operators**:
+
+```python
+In [10]: new_expr = expr & (7 > 10)  # always False
+
+In [11]: new_expr.eval({'x': 7, 'y': 'hello'})
+Out[12]: False
+```
+
+### napari context keys
+
+To capture napari-specific conditions, napari will declare special string
+**names** that can be used in a napari expression.  Taking the example above, a
+plugin might only want to provide a function if “the active layer has at least 3
+dimensions”. For this, napari recognizes the variable `active_layer_ndim` used
+in an expression. In a plugin manifest, the plugin can provide a *when clause*
+to enable/disable a given command:
+
+```yaml
+command:
+   id: myplugin.my_command
+   when: active_layer_ndim >= 3
+```
+
+Some example context key names (currently) include:
+
+| Name     | Description |
+| -------- | -------- |
+| `layers_selection_count`  | Number of layers currently selected  | 
+| `all_layers_linked`  | True when all selected layers are linked  | 
+| `active_layer_is_rgb`  | True when the active layer is RGB  | 
+| `active_layer_type`  | Lowercase name of active layer type, or None if no layer is active.  | 
+| `only_images_selected`  | True when there is at least one selected layer and all selected layers are images  | 
+| `active_layer_ndim` | Number of dimensions in the active layer, or `None` if nothing is active |
+...  many more
+
+
+### `ContextKey` objects
+
+To track the special **name** strings that can be used in expressions, napari
+has the `ContextKey` class.  Instances of `ContextKey` subclass from
+[`ast.Name`](https://docs.python.org/3/library/ast.html#ast.Name) and represent
+a variable name in an expression. Additionally, they have a `description` that
+can be used in documentation, a `default_value`, and a `getter` function that
+can be called to retrieve the current value (the parameters passed to the
+`getter` will depend on the context key... see [Updating
+Contexts](#updating-contexts) below for more).
+
+In order to keep related keys together, `ContextKey` instances will usually be
+declared as class attributes on a `ContextNamespace` class:
+
+```python
+# all of the getters here receive an instance of viewer.layers.selection
+
+class LayerListContextKeys(ContextNamespace):
+    layers_selection_count = ContextKey(
+        default_value=0,
+        description="Number of layers currently selected",
+        getter=lambda sel: len(sel),
+    )
+    active_layer_is_rgb = ContextKey(
+        default_value=False,
+        description="True when the active layer is RGB",
+        getter=lambda sel: getattr(sel.active, "rgb", False),
+    )
+    active_layer_ndim = ContextKey(
+        None,
+        "Number of dimensions in the active layer, or `None` if nothing is active",
+        lambda s: s.active and getattr(s.active.data, "ndim", None),
+    )
+```
+
+You can think of a `ContextNamespace` kind of like an `Enum` class.  And just
```

Comment:
I think I'm missing something, but the members feel more like methods (i.e. functions with names and descriptions that take the same instance as their first argument), than they do enum constants.

Reply:
perhaps that statement isn't very useful if someone doesn't know how an `Enum` is constructed.  the `Enum` class is really just a thin shell on `EnumMeta`, and their members are definitely _not_ constants.  They are _instances_ of the Enum class on which they are declared:

```py
In [1]: from enum import Enum

In [2]: class E(Enum):
   ...:     a = 1
   ...:

In [3]: type(E.a) is E
Out[3]: True
```

(this is very much different than a typical class attribute).  So in a sense, an `Enum` class is more of a namespace for a set of related enum instances:

```py
In [4]: E.__members__
Out[4]: mappingproxy({'a': <E.a: 1>})
```

just as a `ContextNamespace` class is a namespace for a set of ContextKey instances.   perhaps the intricacies of `Enum` make this comparison potentially more confusing that useful?

Reply:
> (i.e. functions with names and descriptions that take the same instance as their first argument)

also, don't think of these members as functions.  Think of them as static `Expr` objects, that just happen to come with a function that, when provided an object, can "grab" the current value for their expression in some specific context

Reply:
In most languages, an enum is just supposed to represent a collection of related named constants. In Python the intent is the same: https://docs.python.org/3/library/enum.html

I don't think anyone should have to understand the current implementation details of Enum in Python to understand the comparison here. So, I think removing the sentence is the best option.